File: pomade-signer-go/mailer/smtp.go
package mailer

import (
	"context"
	"fmt"
	"net/smtp"
	"strings"
)

type SmtpMailer struct {
	Host     string
	Port     int
	User     string
	Password string
}

func (m SmtpMailer) Send(_ context.Context, fromEmail string, fromName string, email Email) error {
	addr := fmt.Sprintf("%s:%d", m.Host, m.Port)

	var auth smtp.Auth
	if m.User != "" {
		auth = smtp.PlainAuth("", m.User, m.Password, m.Host)
	}

	from := fmt.Sprintf("%s <%s>", fromName, fromEmail)
	msg := strings.Join([]string{
		"From: " + from,
		"To: " + email.To,
		"Subject: " + email.Subject,
		"MIME-Version: 1.0",
		`Content-Type: multipart/alternative; boundary="pomade-boundary"`,
		"",
		"--pomade-boundary",
		"Content-Type: text/plain; charset=UTF-8",
		"",
		email.Text,
		"",
		"--pomade-boundary",
		"Content-Type: text/html; charset=UTF-8",
		"",
		email.HTML,
		"",
		"--pomade-boundary--",
	}, "\r\n")

	return smtp.SendMail(addr, auth, fromEmail, []string{email.To}, []byte(msg))
}

File: pomade-signer-rust/src/mailer/smtp.rs
use lettre::{
    AsyncSmtpTransport, AsyncTransport, Message, Tokio1Executor,
    message::{MultiPart, SinglePart, header::ContentType},
    transport::smtp::authentication::Credentials,
};

use crate::mailer::{Email, MailFuture, Mailer};

pub struct SmtpMailer {
    pub host: String,
    pub port: u16,
    pub user: Option<String>,
    pub password: Option<String>,
}

impl Mailer for SmtpMailer {
    fn send(&self, from_email: &str, from_name: &str, email: Email) -> MailFuture {
        let from = format!("{} <{}>", from_name, from_email);
        let host = self.host.clone();
        let port = self.port;
        let credentials = self
            .user
            .as_ref()
            .map(|u| Credentials::new(u.clone(), self.password.clone().unwrap_or_default()));

        Box::pin(async move {
            let message = Message::builder()
                .from(
                    from.parse()
                        .map_err(|e| format!("invalid from address: {e}"))?,
                )
                .to(email
                    .to
                    .parse()
                    .map_err(|e| format!("invalid to address: {e}"))?)
                .subject(&email.subject)
                .multipart(
                    MultiPart::alternative()
                        .singlepart(
                            SinglePart::builder()
                                .header(ContentType::TEXT_PLAIN)
                                .body(email.text),
                        )
                        .singlepart(
                            SinglePart::builder()
                                .header(ContentType::TEXT_HTML)
                                .body(email.html),
                        ),
                )
                .map_err(|e| format!("failed to build message: {e}"))?;

            let mut builder = AsyncSmtpTransport::<Tokio1Executor>::starttls_relay(&host)
                .map_err(|e| format!("smtp relay error: {e}"))?
                .port(port);

            if let Some(creds) = credentials {
                builder = builder.credentials(creds);
            }

            builder
                .build()
                .send(message)
                .await
                .map(|_| ())
                .map_err(|e| format!("smtp send error: {e}"))
        })
    }
}
